File: lib/core/utils/phone_utils.dart
/// Normalizes DRC (+243) phone numbers for WhatsApp, calls and SMS.
class PhoneUtils {
  static const String countryCode = '243';

  static String digitsOnly(String? phone) {
    if (phone == null) return '';
    return phone.replaceAll(RegExp(r'\D'), '');
  }

  /// Returns international digits without "+" (e.g. 243975955375).
  static String normalizeDrc(String? phone) {
    var digits = digitsOnly(phone);
    if (digits.isEmpty) return '';

    // 243243XXXXXXXX → 243XXXXXXXX
    while (digits.startsWith('$countryCode$countryCode') &&
        digits.length > 12) {
      digits = digits.substring(countryCode.length);
    }

    // 2430XXXXXXXXX → 243XXXXXXXXX (0 local après l'indicatif)
    if (digits.startsWith('${countryCode}0') && digits.length > 12) {
      final withoutTrunk = countryCode + digits.substring(countryCode.length + 1);
      if (withoutTrunk.length >= 12) {
        digits = withoutTrunk.length > 12
            ? withoutTrunk.substring(0, 12)
            : withoutTrunk;
      }
    }

    if (digits.startsWith(countryCode) && digits.length >= 11) {
      return digits.length > 12 ? digits.substring(0, 12) : digits;
    }

    if (digits.startsWith('00$countryCode') && digits.length >= 13) {
      digits = digits.substring(2);
      return digits.length > 12 ? digits.substring(0, 12) : digits;
    }

    if (digits.startsWith('0') && digits.length >= 10) {
      return '$countryCode${digits.substring(1)}';
    }

    if (digits.length == 9) {
      return '$countryCode$digits';
    }

    return digits;
  }

  /// wa.me expects digits only, with country code and no "+".
  static String forWhatsApp(String? phone) => normalizeDrc(phone);

  /// First valid WhatsApp number from shop fields (whatsapp, then phone).
  static String? shopWhatsAppNumber({String? whatsapp, String? phone}) {
    for (final raw in [whatsapp, phone]) {
      if (raw == null || raw.trim().isEmpty) continue;
      final normalized = forWhatsApp(raw);
      if (isValidDrc(normalized)) return normalized;
    }
    return null;
  }

  /// tel: URI with leading "+".
  static String forTelUri(String? phone) {
    final normalized = normalizeDrc(phone);
    return normalized.isEmpty ? '' : '+$normalized';
  }

  static String forSms(String? phone) => normalizeDrc(phone);

  static String formatForDisplay(String? phone) {
    final normalized = normalizeDrc(phone);
    if (normalized.length < 12) return phone?.trim() ?? '';
    final local = normalized.substring(3);
    if (local.length == 9) {
      return '+243 ${local.substring(0, 2)} '
          '${local.substring(2, 5)} '
          '${local.substring(5, 7)} '
          '${local.substring(7)}';
    }
    return '+$normalized';
  }

  static bool isValidDrc(String? phone) {
    final normalized = normalizeDrc(phone);
    return normalized.startsWith(countryCode) && normalized.length == 12;
  }

  /// Common formats for matching owner/shop phone fields across local & server.
  static List<String> lookupKeys(String? phone) {
    if (phone == null || phone.trim().isEmpty) return const [];
    final trimmed = phone.trim();
    final normalized = normalizeDrc(phone);
    final keys = <String>{trimmed, normalized, '+$normalized'};
    if (normalized.startsWith(countryCode) && normalized.length >= 12) {
      final local = normalized.substring(3);
      keys.add(local);
      keys.add('0$local');
    }
    keys.removeWhere((k) => k.isEmpty);
    return keys.toList();
  }
}

File: server/api/phone_utils.php
<?php
/**
 * DRC phone normalization for wa.me links (mirrors lib/core/utils/phone_utils.dart).
 */

function phone_digits_only(?string $phone): string
{
    if ($phone === null || trim($phone) === '') {
        return '';
    }
    return preg_replace('/\D/', '', $phone) ?? '';
}

function phone_normalize_drc(?string $phone): string
{
    $cc = '243';
    $digits = phone_digits_only($phone);
    if ($digits === '') {
        return '';
    }

    while (strpos($digits, $cc . $cc) === 0 && strlen($digits) > 12) {
        $digits = substr($digits, strlen($cc));
    }

    if (strpos($digits, $cc . '0') === 0 && strlen($digits) > 12) {
        $withoutTrunk = $cc . substr($digits, strlen($cc) + 1);
        if (strlen($withoutTrunk) >= 12) {
            $digits = strlen($withoutTrunk) > 12
                ? substr($withoutTrunk, 0, 12)
                : $withoutTrunk;
        }
    }

    if (strpos($digits, $cc) === 0 && strlen($digits) >= 11) {
        return strlen($digits) > 12 ? substr($digits, 0, 12) : $digits;
    }

    if (strpos($digits, '00' . $cc) === 0 && strlen($digits) >= 13) {
        $digits = substr($digits, 2);
        return strlen($digits) > 12 ? substr($digits, 0, 12) : $digits;
    }

    if (strlen($digits) > 0 && $digits[0] === '0' && strlen($digits) >= 10) {
        return $cc . substr($digits, 1);
    }

    if (strlen($digits) === 9) {
        return $cc . $digits;
    }

    return $digits;
}

function phone_is_valid_drc(?string $phone): bool
{
    $normalized = phone_normalize_drc($phone);
    return strpos($normalized, '243') === 0 && strlen($normalized) === 12;
}

/** Common formats for matching owner_id across local & server (mirrors PhoneUtils.lookupKeys). */
function phone_lookup_keys(?string $phone): array
{
    if ($phone === null || trim($phone) === '') {
        return [];
    }
    $trimmed = trim($phone);
    $normalized = phone_normalize_drc($phone);
    $keys = array_unique(array_filter([
        $trimmed,
        $normalized,
        $normalized !== '' ? '+' . $normalized : '',
    ]));
    if ($normalized !== '' && strpos($normalized, '243') === 0 && strlen($normalized) >= 12) {
        $local = substr($normalized, 3);
        $keys[] = $local;
        $keys[] = '0' . $local;
    }
    return array_values(array_unique(array_filter($keys)));
}

function shop_whatsapp_number(?string $whatsapp, ?string $phone): ?string
{
    foreach ([$whatsapp, $phone] as $raw) {
        if ($raw === null || trim($raw) === '') {
            continue;
        }
        $normalized = phone_normalize_drc($raw);
        if (phone_is_valid_drc($normalized)) {
            return $normalized;
        }
    }
    return null;
}
